src/main/java/com/amazon/ion/impl/bin/IonRawBinaryWriter_1_1.kt

```diff
@@ -0,0 +1,898 @@
+        val stackIterator: ListIterator<ContainerInfo> = containerStack.iterator()
+        // Walk down the stack until we find an ancestor which already has a patch point
+        while (stackIterator.hasNext() && stackIterator.next().patchPoint == null) {
+            // Logic happens in the loop condition.
+        }
+
+        // The iterator cursor is now positioned on an ancestor container that has a patch point
+        // Ascend back up the stack, fixing the ancestors which need a patch point assigned before us
+        while (stackIterator.hasPrevious()) {
+            val ancestor = stackIterator.previous()
+            if (ancestor.patchPoint == null) {
+                ancestor.patchPoint = patchPoints.pushAndGet { it.clear() }
+            }
+        }
+    }
+
+    override fun writeTaggedPlaceholder() {
+        writeScalar {
+            buffer.writeByte(OpCode.TAGGED_PLACEHOLDER)
+            1
+        }
+    }
+
+    override fun writeTaggedPlaceholderWithDefault(default: Consumer<IonRawWriter_1_1>) {
+        buffer.writeByte(OpCode.TAGGED_PLACEHOLDER_WITH_DEFAULT)
+        currentContainer.length++
+        default.accept(this)
+    }
+
+    override fun writeTaglessPlaceholder(taglessEncodingOpcode: Int) = writeScalar {
+        buffer.write2Bytes(OpCode.TAGLESS_PLACEHOLDER.toByte(), taglessEncodingOpcode.toByte())
+        2
+    }
+
+    override fun stepInDirective(directiveOpcode: Int) {
+        currentContainer = containerStack.push { it.reset(DIRECTIVE, buffer.position()) }
+        buffer.writeByte(directiveOpcode)
+    }
+
+    override fun stepInTaglessElementList(taglessEncodingOpcode: Int) {
+        currentContainer = containerStack.push { it.reset(TE_LIST, buffer.position()) }
+        buffer.write2Bytes(OpCode.TAGLESS_ELEMENT_LIST.toByte(), taglessEncodingOpcode.toByte())
+        currentContainer.metadataOffset++
+        buffer.reserve(lengthPrefixPreallocation)
+    }
+
+    override fun stepInTaglessElementList(macroId: Int, macroName: String?) {
+        currentContainer = containerStack.push { it.reset(TE_LIST, buffer.position()) }
+        buffer.writeByte(OpCode.TAGLESS_ELEMENT_LIST)
+        currentContainer.metadataOffset += writeEExpMacroIdWithoutLengthPrefix(macroId)
+        buffer.reserve(lengthPrefixPreallocation)
+    }
+
+    override fun stepInTaglessElementSExp(taglessEncodingOpcode: Int) {
+        currentContainer = containerStack.push { it.reset(TE_SEXP, buffer.position()) }
+        buffer.write2Bytes(OpCode.TAGLESS_ELEMENT_SEXP.toByte(), taglessEncodingOpcode.toByte())
+        currentContainer.metadataOffset++
+        buffer.reserve(lengthPrefixPreallocation)
+    }
+
+    override fun stepInTaglessElementSExp(macroId: Int, macroName: String?) {
+        currentContainer = containerStack.push { it.reset(TE_SEXP, buffer.position()) }
+        buffer.writeByte(OpCode.TAGLESS_ELEMENT_SEXP)
+        currentContainer.metadataOffset += writeEExpMacroIdWithoutLengthPrefix(macroId)
```

Comment:
Do we need a variant that allows the macro shapes to be length-prefixed? (We have an example of this in the spec)